b_cfn_elasticsearch_cloner/resource.py

```diff
@@ -0,0 +1,211 @@
+from typing import Optional
+
+from aws_cdk.core import Construct, Duration, CustomResource, RemovalPolicy
+from aws_cdk.aws_dynamodb import Table
+from aws_cdk.aws_lambda import Function, SingletonFunction, Runtime, StartingPosition, Code
+from aws_cdk.aws_lambda_event_sources import DynamoEventSource
+from aws_cdk.aws_logs import RetentionDays
+from aws_cdk.aws_iam import Role, ServicePrincipal, PolicyStatement, PolicyDocument, Effect
+from aws_cdk.aws_kms import Key
+
+from b_cfn_elasticsearch_index.resource import ElasticsearchIndexResource
+from b_elasticsearch_layer.layer import Layer as ElasticsearchLayer
+
+from b_cfn_elasticsearch_cloner.cloner_source import root as cloner_root
+from b_cfn_elasticsearch_cloner.initial_cloner_source import root as initial_cloner_root
+
+
+class ElasticsearchCloner(Construct):
+    """
+    Custom resource used for managing an Elasticsearch cloner.
+
+    Creates an index on stack creation.
+    Updates the index on settings change.
+    Deletes the index on stack deletion.
+    """
+
+    def __init__(
+        self,
+        scope: Construct,
+        id: str,
+        elasticsearch_index: ElasticsearchIndexResource,
+        dynamodb_table: Table,
+        kms_key: Optional[Key],
+    ) -> None:
+        super().__init__(scope=scope, id=id)
+
+        initial_cloner_function = SingletonFunction(
+            scope=self,
+            id="InitialClonerFunction",
+            uuid="e01116a4-f939-43f2-8f5b-cc9f862c9e01",
+            lambda_purpose="InitialClonerSingletonLambda",
+            code=Code.from_asset(initial_cloner_root),
+            handler="index.handler",
+            runtime=Runtime.PYTHON_3_8,
+            layers=[ElasticsearchLayer],
+            log_retention=RetentionDays.ONE_MONTH,
+            memory_size=128,
+            role=Role(
+                scope=self,
+                id="InitialClonerFunctionRole",
```

Comment:
Do not hardcode ids. Two same resources would result in a CDK error. Instead of taking an "id" in the constructor, take a prefix and add that prefix to all of your resource names and ids.